### ISSUE TITLE
foreach pass-by-reference leads to "Avoid using undefined variables such as '$item' which will lead to PHP notices" error

### ISSUE THREAD

**Issue by yutv** [phpmd/phpmd#720]
- PHPMD version: 2.8.1, 2.8.0, 2.8.0-beta.2, 2.8.0-beta.1
- PHP Version: 7.2.26
- Installation type: composer
- Operating System / Distribution & Version: Ubuntu 18.04.3 LTS

### Current Behavior
Foo.php
```php
<?php

class Foo
{
    public function bar(array $array)
    {
        foreach ($array as &$item) {
            $item++;
        }
        return $array;
    }
}
```
**Result:**

    $ vendor/bin/phpmd Foo.php text cleancode && echo "Success"
    /home/user/project/Foo.php:7  Avoid using undefined variables such as '$item' which will lead to PHP notices.
    /home/user/project/Foo.php:8  Avoid using undefined variables such as '$item' which will lead to PHP notices.

### Expected Behavior

    $ vendor/bin/phpmd Foo.php text cleancode && echo "Success"
    Success

### Steps To Reproduce:
See the example above

## Checks before submitting
* [x] Be sure that there isn't already an issue about this. See: [Issues list](https://github.com/phpmd/phpmd/issues)
* [x] Be sure that there isn't already a pull request about this. See: [Pull requests](https://github.com/phpmd/phpmd/pulls)
* [x] I have added every step to reproduce the bug.
* [x] If possible I added relevant code examples.
* [x] This issue is about 1 bug and nothing more.
* [x] The issue has a descriptive title. For example:  "JSON rendering failed on Windows for filenames with space".

**Comment by char101** [comment]
Patch for this bug

```diff
diff --git a/src/main/php/PHPMD/Rule/CleanCode/UndefinedVariable.php b/src/main/php/PHPMD/Rule/CleanCode/UndefinedVariable.php
index dd90665..110daaf 100644
--- a/src/main/php/PHPMD/Rule/CleanCode/UndefinedVariable.php
+++ b/src/main/php/PHPMD/Rule/CleanCode/UndefinedVariable.php
@@ -18,6 +18,7 @@
 namespace PHPMD\Rule\CleanCode;

 use PDepend\Source\AST\ASTVariable;
+use PDepend\Source\AST\ASTUnaryExpression;
 use PDepend\Source\AST\State;
 use PHPMD\AbstractNode;
 use PHPMD\Node\AbstractCallableNode;
@@ -138,6 +139,10 @@ class UndefinedVariable extends AbstractLocalVariable implements FunctionAware,
             foreach ($foreachStatement->getChildren() as $children) {
                 if ($children instanceof ASTVariable) {
                     $this->addVariableDefinition($children);
+                } elseif ($children instanceof ASTUnaryExpression) {
+                    foreach ($children->getChildren() as $refChildren) {
+                        $this->addVariableDefinition($refChildren);
+                    }
                 }
             }
         }
```

**Comment by kylekatarnls** [comment]
Awesome, this could be a very good alternative until it get fixed in PDepend.

**Comment by char101** [comment]
Since I have too many forks already I prefer if you could apply the patch yourself.

**Comment by kylekatarnls** [comment]
No problem I will add it and few tests.

**Comment by kylekatarnls** [comment]
Merged, you can test it requiring `"phpmd/phpmd": "dev-master"` it will be released in 2.8.3.

**Comment by bkdotcom** [comment]
This is still an issue in 2.9.0

```php
foreach ($data as $k => &$row) {   // error triggered here
    $row['foo'] = 'bar';                         // and here
}
```
